SMASH Integration

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1883 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->
- d790880de0392f22e8d2d2d4096e1d44c3e60616
  :round_pushpin: **allow different ways of building metadata URL from metadata information**
  The idea is the following:

- We have a list of "URL builders" that will be tried in order, but only
  if the first builder replied with a connection failure (so, only if the
  server is somewhat down)

- We can always have our current behavior by using the
  'identityUrlBuilder'.

- Then, we can augment this behavior if given a specific registry URL
  via the command-line options, which will add a "registryUrlBuilder"
  which takes predecence over the identity builder.

- 48884cd3a0c436b0e4b4e817e8df2c7319440374
  :round_pushpin: **try fetching metadata from the next builder on HTTP exception**
  
- c133609672d5b1173a00a20595bc00ac9054ee10
  :round_pushpin: **add command-line option for enabling a pool metadata proxy**
  Going for a command-line option so that this can easily be turned on and off at will in case of any issue. Plus, not everyone might agree with the idea of going through a registry so hard-coding the registry URL is a no go. cardano-wallet-jormungandr uses an ad-hoc environment variable for this, but a documented option in the command-line is definitely cleaner.

- e9dbf56e160b869af3b677e82449e56b8b8a37af
  :round_pushpin: **revise logging around pool metadata fetching to have more insight _inside_ the fetcher**
  So that we can clearly log fallback attempts and keep it clean.

- cf00db653779cf80a579025e8df8a9576c14b46a
  :round_pushpin: **Regenerate nix**
  
- 08305897348945dd16afd484d2b500fd2b298e52
  :round_pushpin: **use 'Either HttpException' as a base monad instead of 'ExceptT'**
    The underlying URL builder relies on the 'MonadThrow' class to send errors. Yet, when using 'ExceptT', the error is thrown _in the base monad_!!
  So, using 'Either' instead in conjunction with `except` to catch the error as expected in the loopl, instead of throwing hard in the parent context.

- dd00d12e38defb608db5565f181c0052c3713e44
  :round_pushpin: **be more specific when interpreting responses from the aggregation server**
  
- 28b9d575c23413047b73819aa60afb3bb65d6427
  :round_pushpin: **remove fallback when a pool metadata proxy is supplied.**
  The fallback made sense in the early days, but we can now make it a proper switch: it's either fetching from pools, or it's fetching from the registry


# Comments

<!-- Additional comments or screenshots to attach if any -->

- [ ] :warning: TODO: show that it works. So far, on the shelley testnet:

Some quick manual / sanity testing:

```console
$ cardano-wallet-shelley serve \
  --node-socket node.socket \
  --testnet shelley_testnet-genesis.json \
  --pool-metadata-proxy https://smash.shelley-testnet.dev.cardano.org/api/v1/metadata

[...]

[2020-07-08 19:57:09.29 UTC] Fetching metadata with hash c9ed7096 from https://smash.shelley-testnet.dev.cardano.orgapi/v1/metadata/c9ed70969c964284ff854d61e43c2591b67ceb7e978347cb1b50ba3ed1d3edb4
[2020-07-08 19:57:09.45 UTC] Successfully fetched metadata with hash c9ed7096: StakePoolMetadata {ticker = StakePoolTicker {unStakePoolTicker = "IOHK1"}, name = "IOHK1", description = Just "IOHK Pool", homepage = "https://iohk.io"}
```

```console
$ cardano-wallet-shelley serve \
  --node-socket node.socket \
  --testnet shelley_testnet-genesis.json \
  --pool-metadata-proxy patate
option --pool-metadata-proxy: Invalid URI. Make sure the URI is well formed, with a protocol and a host.
```

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
allow different ways of building metadata URL from metadata information

The idea is the following:

- We have a list of "URL builders" that will be tried in order, but only
  if the first builder replied with a connection failure (so, only if the
  server is somewhat down)

- We can always have our current behavior by using the
  'identityUrlBuilder'.

- Then, we can augment this behavior if given a specific registry URL
  via the command-line options, which will add a "registryUrlBuilder"
  which takes predecence over the identity builder.

KtorZ committed:
try fetching metadata from the next builder on HTTP exception

KtorZ committed:
add command-line option for enabling a pool metadata proxy

Going for a command-line option so that this can easily be turned on and off at will in case of any issue. Plus, not everyone might agree with the idea of going through a registry so hard-coding the registry URL is a no go. cardano-wallet-jormungandr uses an ad-hoc environment variable for this, but a documented option in the command-line is definitely cleaner.

KtorZ committed:
revise logging around pool metadata fetching to have more insight _inside_ the fetcher

So that we can clearly log fallback attempts and keep it clean.

piotr-iohk commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -70,45 +104,129 @@ defaultManagerSettings =
+            -- to be less than 512 bytes. For security reasons, we only download the
+            -- first 512 bytes.
+            if responseStatus res /= status200 then
+                return $ Left "Server did reply, but not positively. This metadata may be blacklisted."
```
Does SMASH respond with specific code when metadata is "blacklisted"? If so maybe it could be specified for this particular case.

Currently the solution falls back only if it SMASH url cannot be reached, but for example if provided SMASH url returns `404` it could be perfectly falling back as well...

KtorZ replied:
The blacklisting isn't yet implemented in SMASH but, when it does it'll probably return a `404` (according to the specification). There really are two cases:

a) The metadata aren't yet available on SMASH because it hasn't yet downloaded the metadata from a pool (that could be the case for recent pools). Then, it's okay, we'll get a `404` and will retry the url _later_ (exponential backoff), up to several days after the first try and up to 10 times.

b) The metadata is blacklisted, in which case SMASH should also reply with a 404 (very much as if the metadata was absent), and here, we'll retry all-the-same because we can't actually tell the difference with the first case. After 10 retries, we'll stop and consider this metadata "unfetchable". 

When blacklisting is implemented, we might revise this behavior and "acknowledge" blacklisted pools if possible so that we can skip the 10 retries.

piotr-iohk replied:
Then maybe we can be more specific than `/= status200`, and just implement this behaviour on `== status404`? I can imagine that in case of mainetenace, the server might for example return `502` or something else in which case we could just fall back.

piotr-iohk replied:
Also as far as I see SMASH actually responds with 200 and a message, in case of data missing:

```
{"code":"DbLookupTxMetadataHash","description":"The hash 534b6a6bd10d23a2eb060a51d61df263bf2fa3f10b6f5a701fd9a77d6c81e67011111111111111111111 is missing from the DB."}
```

KtorZ replied:
> Also as far as I see SMASH actually responds with 200 and a message, in case of data missing:

:facepalm: really?

KtorZ replied:
https://github.com/input-output-hk/smash/issues/16

KtorZ committed:
use 'Either HttpException' as a base monad instead of 'ExceptT'

  The underlying URL builder relies on the 'MonadThrow' class to send errors. Yet, when using 'ExceptT', the error is thrown _in the base monad_!!
  So, using 'Either' instead in conjunction with `except` to catch the error as expected in the loopl, instead of throwing hard in the parent context.

KtorZ committed:
be more specific when interpreting responses from the aggregation server

KtorZ committed:
remove fallback when a pool metadata proxy is supplied.

The fallback made sense in the early days, but we can now make it a proper switch: it's either fetching from pools, or it's fetching from the registry

rvl commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -70,45 +102,133 @@ defaultManagerSettings =
+        { uriPath = "/" <> intercalate "/" (pathSegments baseUrl ++ [hash])
+        }
+  where
+    hash = T.unpack $ T.decodeUtf8 $ convertToBase Base16 bytes
```
```suggestion
    hash = B8.unpack $ convertToBase Base16 bytes
```

rvl commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -70,45 +102,133 @@ defaultManagerSettings =
+    -- Try each builder in order, but only if the previous builder led to a
+    -- "ConnectionFailure" exception. Other exceptions mean that we could
+    -- likely reach the server.
+    fromFirst (builder:rest) = do
```
I found this a little confusing. Can we separate the function which makes a request and reads at most 512 bytes from the response from the function which tries each builder and uses the first which succeeds?

KtorZ replied:
I can try that :+1:

rvl commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1296,6 +1299,21 @@ tlsOption = TlsConfiguration
+poolMetadataProxyOption
+    :: Parser URI
+poolMetadataProxyOption = option (eitherReader reader) $ mempty
+    <> long "pool-metadata-proxy"
```
I would prefer to not use the word "proxy" where it could be confused with actual HTTP proxies.
We could just say `--smash-server`.
And what about making the CF smash server the default, with an option to disable the metadata aggregator?

KtorZ replied:
> I would prefer to not use the word "proxy" where it could be confused with actual HTTP proxies.

Fair point. Although I dislike the "--smash-server" naming for this because in principle, smash is one implementation of a metadata registry. There could be others by the community that aren't smash. Hence why I wanted to keep the name rather "generic". Perhaps `--pool-metadata-registry` ?

>  And what about making the CF smash server the default, with an option to disable the metadata aggregator?

There's no CF server yet as far as I know :grimacing: ... plus, with all the networks and candidates at the moment, it's sort of hard to know what a sensible default would be (other than, no default). Long-term, I do however agree, once the dust settle down a bit.

rvl replied:
I thought "smash" _was_ a generic name, or at least something that implements the smash API.
I have asked for clarification input-output-hk/smash#22.

KtorZ replied:
Anyway, I'll for "--smash-server" since it's more transparent for now. We can always change later.

KtorZ committed:
split 'getChunk' from 'fromFirst', separating the fetching logic from the retrying plumbing.

rvl committed:
Fix tests/bench compile

rvl commented on lib/core/src/Cardano/Pool/Metadata.hs:
```diff
@@ -70,45 +102,138 @@ defaultManagerSettings =
   where
-    StakePoolMetadataUrl  url  = url_
-    StakePoolMetadataHash hash = hash_
+    runExceptTLog
+        :: ExceptT String IO StakePoolMetadata
+        -> IO (Maybe StakePoolMetadata)
+    runExceptTLog action = runExceptT action >>= \case
+        Left msg ->
+            Nothing <$ traceWith tr (MsgFetchPoolMetadataFailure hash msg)
+
+        Right meta ->
+            Just meta <$ traceWith tr (MsgFetchPoolMetadataSuccess hash meta)
```
Previously I have combined Success/Failure log messages into a single message type, e.g.
```
    | MsgFetchPoolMetadataResult StakePoolMetadataHash (Either String StakePoolMetadata)
```
This way is no less expressive for logging, and means you can use a single `traceWith` statement and avoid needing to use constructs like `runExceptTLog`.

rvl committed:
Change --pool-metadata-proxy to --smash-url

rvl committed:
Regenerate nix

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1296,6 +1299,26 @@ tlsOption = TlsConfiguration
+    err :: String
+    err =
+        "Invalid URI. Make sure it is a well-formed \
+        \ absolute http or https URI."
```
:+1:

KtorZ commented on lib/shelley/test/integration/Main.hs:
```diff
@@ -165,9 +165,7 @@ testPoolConfigs =
-    , PoolConfig {retirementEpoch = Just 3}
-      -- This pool should retire, but not within the duration of a test run:
-    , PoolConfig {retirementEpoch = Just 100_000}
+    , PoolConfig {retirementEpoch = Just 1_000}
```
Arf! I think there has been a wrong rebase / conflict resolution here! Hence the integration failure.

rvl replied:
Oops
